src/core/patch.rkt

```diff
@@ -39,24 +39,32 @@
         ; if disabled, only implementation selection
         `((,lowering-rules . ((iteration . 1) (scheduler . simple))))))
 
+  (define roots
+    (for/vector ([approx (in-list approxs)])
+      (batchref-idx (alt-expr approx))))
+
   ; run egg
-  (define batch (progs->batch (map alt-expr approxs)))
-  (define runner (make-egg-runner batch (batch-roots batch) reprs schedule))
+  (define runner (make-egg-runner global-batch roots reprs schedule))
   (define simplification-options
     (simplify-batch runner
-                    (typed-egg-extractor
-                     (if (*egraph-platform-cost*) platform-egg-cost-proc default-egg-cost-proc))))
+                    (typed-egg-batch-extractor
+                     (if (*egraph-platform-cost*) platform-egg-cost-proc default-egg-cost-proc)
+                     global-batch)))
 
   ; convert to altns
   (define simplified
     (reap [sow]
+          (define global-batch-mutable (batch->mutable-batch global-batch)) ; Create mutable batch
```

Comment:
I don't get why the global-batch has to be stored as an immutable batch? Why not just make it a mutable batch the whole time, and then we don't convert between them?

Reply:
Yeah, that makes sense. Probably everything can be done with a `mutable-batch`, just some hacks will be needed to implement, so that extraction and input to `egg` will also be working with `mutable-batch`

Reply:
Update: it can not be done trivially everywhere, because sometimes we also need to access nodes in the batch.
For example when we run `egg` we need to take inputs for the batch which means accessing nodes by indexes.
`batch` at this points will be faster to access rather than `mutable-batch` since `batch` uses vector as nodes, while `mutable-batch` uses list.
The difference in `list-ref` and `vector-ref` is about 5 times.
Here is a code that measures it:
```> (define l (build-list 10000 values))
> (define v (build-vector 10000 values))
> (time (for ([i (in-range 10000)])
          (list-ref l i)))
cpu time: 91 real time: 92 gc time: 0
> (time (for ([i (in-range 10000)])
          (vector-ref v i)))
cpu time: 19 real time: 19 gc time: 0
```

Reply:
However, while keeping `batch` and `mutable-batch` around we also introduce some overheads, particularly on every conversion between them.
`batch->mutable-batch`'s overhead is restoring the index, for nodes length `10000` it takes approximately:
```
(time (make-hash (for/list ([node (in-list l)]
                        [n (in-naturals)])
               (cons node n))))
cpu time: 19 real time: 19 gc time: 0
```
While `mutable-batch->batch` takes almost nothing:
```
(time (list->vector (reverse l)))
cpu time: 0 real time: 0 gc time: 0
```

Reply:
It seems that we may be still winning with `batch` and `mutable-batch` being around due to the saving time of node-access